cmd/sharded-test-server/virtual.go

```diff
@@ -27,30 +27,106 @@ import (
 
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/cmd/sharded-test-server/third_party/library-go/crypto"
 	"github.com/kcp-dev/kcp/cmd/test-server/helpers"
 	"github.com/kcp-dev/kcp/test/e2e/framework"
 )
 
-func startVirtual(ctx context.Context, index int, logDirPath, workDirPath string) (<-chan error, error) {
+func startVirtual(ctx context.Context, index int, servingCA *crypto.CA, hostIP string, logDirPath, workDirPath string, clientCA *crypto.CA) (<-chan error, error) {
+	logger := klog.FromContext(ctx)
+
 	prefix := fmt.Sprintf("VW-%d", index)
 	yellow := color.New(color.BgYellow, color.FgHiWhite).SprintFunc()
 	out := lineprefix.New(
 		lineprefix.Prefix(yellow(prefix)),
 		lineprefix.Color(color.New(color.FgHiYellow)),
 	)
 
+	// create serving cert
+	hostnames := sets.NewString("localhost", hostIP)
+	logger.Info("Creating vw server serving cert", "index", index, "hostnames", hostnames.List())
+	cert, err := servingCA.MakeServerCert(hostnames, 365)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create server cert: %w", err)
+	}
+	servingKeyFile := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/apiserver.key", index))
+	servingCertFile := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/apiserver.crt", index))
+	if err := cert.WriteCertConfigFile(servingCertFile, servingKeyFile); err != nil {
+		return nil, fmt.Errorf("failed to write server cert: %w", err)
+	}
+
+	// create client cert used to talk to kcp
+	vwClientCert := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/shard-client-cert.crt", index))
+	vwClientCertKey := filepath.Join(workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/shard-client-cert.key", index))
+	shardUser := &user.DefaultInfo{Name: fmt.Sprintf("kcp-vw-%d", index), Groups: []string{"system:masters"}}
+	_, err = clientCA.MakeClientCertificate(vwClientCert, vwClientCertKey, shardUser, 365)
+	if err != nil {
+		fmt.Printf("failed to create vw client cert: %v\n", err)
+		os.Exit(1)
+	}
+
+	servingCAPath, err := filepath.Abs(filepath.Join(workDirPath, ".kcp/serving-ca.crt"))
+	if err != nil {
+		fmt.Printf("error getting absolute path for %q: %v\n", filepath.Join(workDirPath, ".kcp/serving-ca.crt"), err)
+		os.Exit(1)
+	}
+	vwClientCertPath, err := filepath.Abs(vwClientCert)
+	if err != nil {
+		fmt.Printf("error getting absolute path for %q: %v\n", vwClientCert, err)
+		os.Exit(1)
+	}
+	vwClientCertKeyPath, err := filepath.Abs(vwClientCertKey)
+	if err != nil {
+		fmt.Printf("error getting absolute path for %q: %v\n", vwClientCertKey, err)
+		os.Exit(1)
+	}
+
+	virtualWorkspaceKubeConfig := clientcmdapi.Config{
+		Clusters: map[string]*clientcmdapi.Cluster{
+			"shard": {
+				Server:               fmt.Sprintf("https://localhost:%d/clusters/system:admin", 6444+index),
```

Comment:
@ncdc should this be `7444+index` so the `virtualworkspace.kubeconfig` points directly at the VW endpoint?

Reply:
context - I realize this will redirect to the standalone VW server, but I was planning to add a readiness check for the VW readyz, and I realized I can't use this kubeconfig like we do e.g for [the cache server](https://github.com/kcp-dev/kcp/blob/main/cmd/sharded-test-server/cache.go#L92)